File: server/Broadsock.cpp
/*
 * Based on https://github.com/yorickdewid/Chat-Server
 * And http://www.geeksforgeeks.org/socket-programming-in-cc-handling-multiple-clients-on-server-without-multi-threading/
 */

#include <Broadsock.h>
#include <Message.h>
#include <sys/socket.h>
#include <netinet/in.h>
#include <arpa/inet.h>
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <errno.h>
#include <string.h>
#include <sys/types.h>

#define TRUE	1
#define FALSE	0


Broadsock::Broadsock() {
	for(int i = 0; i < MAX_CLIENTS; i++) {
		clients[i] = 0;
	}
	clientCount = 0;
	uid = 10;
}

/* Add client to queue */
void Broadsock::QueueAdd(Client *client) {
	for(int i = 0; i < MAX_CLIENTS; i++) {
		if(!clients[i]) {
			clients[i] = client;
			clientCount++;
			return;
		}
	}
}

/* Delete client from queue */
void Broadsock::QueueDelete(int uid) {
	for(int i = 0; i < MAX_CLIENTS; i++) {
		if(clients[i]) {
			if(clients[i]->uid == uid) {
				clients[i] = NULL;
				clientCount--;
				return;
			}
		}
	}
}

bool Broadsock::IsEmpty() {
	return clientCount == 0;
}

void Broadsock::Send(int fd, Message message) {
	int length = message.MessageLength();
	char* bytes = message.MessageContent();
	unsigned char b0 = (length & 0xFF000000) >> 24;
	unsigned char b1 = (length & 0x00FF0000) >> 16;
	unsigned char b2 = (length & 0x0000FF00) >> 8;
	unsigned char b3 = (length & 0x000000FF);
	send(fd, &b0, 1, 0);
	send(fd, &b1, 1, 0);
	send(fd, &b2, 1, 0);
	send(fd, &b3, 1, 0);
	send(fd, bytes, length, 0);
}

/* Send message to all clients but the sender */
void Broadsock::SendMessage(Message message, int uid) {
	for(int i = 0; i < MAX_CLIENTS; i++) {
		Client* client = clients[i];
		if(client) {
			if(client->uid != uid) {
				Send(client->connfd, message);
			}
		}
	}
}

/* Send message to all clients */
void Broadsock::SendMessageAll(Message message) {
	for(int i=0;i<MAX_CLIENTS;i++) {
		Client* client = clients[i];
		if(client) {
			Send(client->connfd, message);
		}
	}
}

/* Send message to sender */
void Broadsock::SendMessageSelf(Message message, int connfd) {
	for(int i = 0; i < MAX_CLIENTS; i++) {
		Client* client = clients[i];
		if(client) {
			if(client->connfd == connfd) {
				Send(client->connfd, message);
				return;
			}
		}
	}
}

/* Send message to client */
void Broadsock::SendMessageClient(Message message, int uid) {
	for(int i = 0; i < MAX_CLIENTS; i++) {
		Client* client = clients[i];
		if(client) {
			if(client->uid == uid) {
				Send(client->connfd, message);
				return;
			}
		}
	}
}

/**
 * Handle a disconnected client
 * This will remove the client from the queue and yield thread
 */
void Broadsock::HandleClientDisconnected(Client* client) {

	// Notify connected clients of the disconnect
	Message disconnectMessage;
	disconnectMessage.WriteNumber(client->uid);
	disconnectMessage.WriteNullString("DISCONNECT\0");
	SendMessage(disconnectMessage, client->uid);

	printf("<<DISCONNECT %s:%d REFERENCED BY %d\n", inet_ntoa(client->addr.sin_addr), ntohs(client->addr.sin_port), client->uid);

	// Close connection
	close(client->connfd);

	// Delete client from queue
	QueueDelete(client->uid);
	free(client);

	if (clientCount == 0) {
		//TerminateGameSession() ?
		printf("All clients have disconnected\n");
	}
}

/**
 * Handle a connected client
 * This will create a Client struct and notify other clients of the connection
 * The connected client will also receive a message with the assigned user id
 */
bool Broadsock::HandleClientConnected(struct sockaddr_in client_addr, int connfd) {
	// Client settings
	Client *client = (Client *)malloc(sizeof(Client));
	client->addr = client_addr;
	client->connfd = connfd;
	client->customData[0] = 0;
	client->uid = uid++;

	printf("<<CONNECT %s:%d REFERENCED BY %d\n", inet_ntoa(client->addr.sin_addr), ntohs(client->addr.sin_port), client->uid);

	// Add client to the queue
	QueueAdd(client);

	// Notify all clients of the new client
	Message connectOtherMessage;
	connectOtherMessage.WriteNumber(client->uid);
	connectOtherMessage.WriteNullString("CONNECT_OTHER\0");
	connectOtherMessage.WriteNullString(inet_ntoa(client_addr.sin_addr));
	connectOtherMessage.WriteNumber(client->addr.sin_port);
	printf("Sending connectOtherMessage message\n");
	SendMessage(connectOtherMessage, client->uid);

	// Notify self of uid
	Message connectSelfMessage;
	connectSelfMessage.WriteNumber(client->uid);
	connectSelfMessage.WriteNullString("CONNECT_SELF\0");
	printf("Sending connectSelfMessage message\n");
	SendMessageSelf(connectSelfMessage, client->connfd);
	return true;
}


void Broadsock::HandleClientMessage(Client* client, Message message) {
	// Add sender id to the message and broadcast it
	Message out;
	out.WriteNumber(client->uid);
	out.WriteBytes(message.MessageContent(), message.MessageLength());
	SendMessage(out, client->uid);
}


bool Broadsock::Connect() {
	// Create master socket
	if((masterfd = socket(AF_INET, SOCK_STREAM, 0)) == 0) {
		perror("Socket creation failed");
		return false;
	}

	// Set master socket to allow multiple connections
	int opt = TRUE;
	if(setsockopt(masterfd, SOL_SOCKET, SO_REUSEADDR, (char *)&opt, sizeof(opt)) < 0)
	{
		perror("Socket setsockopt failed");
		return false;
	}

	// Socket settings
	struct sockaddr_in serverAddr;
	serverAddr.sin_family = AF_INET;
	serverAddr.sin_addr.s_addr = htonl(INADDR_ANY);
	serverAddr.sin_port = htons(PORT);

	// Bind master socket
	if(bind(masterfd, (struct sockaddr*)&serverAddr, sizeof(serverAddr)) < 0) {
		perror("Socket binding failed");
		return false;
	}

	// Listen on master socket
	if(listen(masterfd, 10) < 0) {
		perror("Socket listening failed");
		return false;
	}

	printf("<[SERVER STARTED]>\n");

	return true;
}

bool Broadsock::Start() {
	// Accept clients
	fd_set readfds;
	while (TRUE) {
		// Clear the socket set
		FD_ZERO(&readfds);

		//add master socket to set
		FD_SET(masterfd, &readfds);

		// Highest file descriptor number, need it for the select function
		int max_sd = masterfd;

		//add child sockets to set
		for(int i = 0; i < MAX_CLIENTS; i++) {
			Client* client = clients[i];
			if(client) {
				FD_SET(client->connfd, &readfds);
				if(client->connfd > max_sd) {
					max_sd = client->connfd;
				}
			}
		}

		// Wait for an activity on one of the sockets, timeout is NULL,
		// so wait indefinitely
		int activity = select(max_sd + 1 , &readfds , NULL , NULL , NULL);
		if((activity < 0) && (errno != EINTR)) {
			printf("select error\n");
		}

		// If something happened on the master socket,
		// then its an incoming connection
		if (FD_ISSET(masterfd, &readfds))
		{
			struct sockaddr_in address;
			socklen_t addrlen = sizeof(address);
			int connfd;
			if ((connfd = accept(masterfd, (struct sockaddr *)&address, (socklen_t*)&addrlen)) < 0) {
				perror("accept");
				exit(EXIT_FAILURE);
			}

			// Check if max clients is reached
			if ((clientCount + 1) == MAX_CLIENTS) {
				printf("<<REJECT %s:%d (MAX CLIENTS REACHED)\n", inet_ntoa(address.sin_addr), ntohs(address.sin_port));
				close(connfd);
			}
			else {
				HandleClientConnected(address, connfd);
			}
		}

		// Check if the clients have any data to read
		for(int i = 0; i < MAX_CLIENTS; i++) {
			Client* client = clients[i];
			if(client) {
				if (FD_ISSET(client->connfd , &readfds)) {
					// Read message length
					unsigned char len[4];
					int ret = read(client->connfd, &len, 4);
					if(ret == 0) {
						HandleClientDisconnected(client);
						continue;
					}
					int length = len[0] << 24 | len[1] << 16 | len[2] << 8 | len[3];

					unsigned char ch;
					// Ignore messages that are too long
					if(length > MESSAGE_SIZE) {
						for(int i=0; i<length; i++) {
							int ret;
							if((ret = read(client->connfd, &ch, 1)) == 0) {
								HandleClientDisconnected(client);
								continue;
							}
						}
						continue;
					}

					// Read the message
					Message message;
					for(int i=0; i<length; i++) {
						int ret;
						if((ret = read(client->connfd, &ch, 1)) == 0) {
							HandleClientDisconnected(client);
							continue;
						}
						message.WriteByte(ch);
					}
					message.Rewind();

					// Handle message
					HandleClientMessage(client, message);
				}
			}
		}
	}
	return true;
}

File: broadsock/server/broadsock.lua
local socket = require "builtins.scripts.socket"
local tcp_writer = require "broadsock.util.tcp_writer"
local tcp_reader = require "broadsock.util.tcp_reader"
local stream = require "broadsock.util.stream"


local M = {}

local clients = {}

local uid_sequence = 0

local server_socket

local function log(...)
	--print("[BROADSOCK SERVER]", ...)
end

local function tomessage(str)
	return stream.number_to_int32(#str) .. str
end

local function add_client(client)
	assert(client)
	log("add_client", client.uid)
	table.insert(clients, client)
end

local function remove_client(uid_to_remove)
	assert(uid_to_remove, "You must provide an uid")
	log("remove_client", uid_to_remove)
	for i=1,#clients do
		local client = clients[i]
		if client.uid == uid_to_remove then
			log("remove_client - removed")
			table.remove(clients, i)
			return
		end
	end
end

function M.client_count()
	return #clients
end

function M.send_message(client, message)
	assert(client, "You must provide a client")
	assert(message, "You must provide a message")
	log("send uid:", client.uid, "message:", message, "length:", #message)
	client.writer.add(message)
	client.writer.send()
end

function M.send_message_others(message, uid)
	assert(message, "You must provide a message")
	assert(uid, "You must provide a uid")
	--log("send_message_others", uid, message)
	for i=1,#clients do
		local client = clients[i]
		if client.uid ~= uid then
			M.send_message(client, message)
		end
	end
end

function M.send_message_all(message)
	assert(message, "You must provide a message")
	log("send_message_all", message)
	for i=1,#clients do
		local client = clients[i]
		M.send_message(client, message)
	end
end

function M.send_message_client(message, uid)
	assert(message, "You must provide a message")
	assert(uid, "You must provide a uid")
	log("send_message_client", uid, message)
	for i=1,#clients do
		local client = clients[i]
		if client.uid == uid then
			M.send_message(client, message)
			break
		end
	end
end

local function create_client(ip, uid, skt, data)
	assert(ip, "You must provide an ip")
	assert(uid, "you must provide an uid")
	assert(skt, "You must provide a socket")
	local client = {}
	client.ip = ip
	client.uid = uid
	client.data = data
	client.socket = skt
	client.writer = tcp_writer.create(skt)
	client.reader = tcp_reader.create(skt, function(message, length)
		M.handle_client_message(client, message)
	end)
	return client
end

function M.handle_client_message(client, message)
	assert(client, "You must provide a client")
	assert(message, "You must provide a message")
	log("handle_client_message", client.uid, stream.dump(message))
	local out = stream.writer().number(client.uid).bytes(message).tostring()
	M.send_message_others(tomessage(out), client.uid)
end

function M.handle_client_disconnected(client)
	assert(client, "You must provide a client")
	local disconnect_message = tomessage(
		stream.writer()
			.number(client.uid)
			.string("DISCONNECT")
			.tostring()
	)
	M.send_message_all(disconnect_message)

	remove_client(client.uid)
end

function M.handle_client_connected(ip, port, skt)
	assert(ip, "You must provide an ip")
	assert(port, "You must provide a port")
	assert(skt, "You must provide a socket")
	uid_sequence = uid_sequence + 1
	local client = create_client(ip, uid_sequence, skt, nil)
	add_client(client)

	local other_message = tomessage(
		stream.writer()
			.number(client.uid)
			.string("CONNECT_OTHER")
			.string(ip)
			.number(port)
			.tostring()
	)
	M.send_message_others(other_message, client.uid)

	local self_message = tomessage(
		stream.writer()
			.number(client.uid)
			.string("CONNECT_SELF")
			.tostring()
	)
	M.send_message(client, self_message)
	return client
end

--- Start the server. This will set up the server socket.
-- @param port The port to listen to connections on
-- @return success True on success, otherwise false
-- @return error Error message on failure, otherwise nil
function M.start(port)
	assert(port, "You must provide a port")
	log("Starting TCP server on port " .. port)
	local ok, err = pcall(function()
		local skt, err = socket.bind("*", port)
		assert(skt, err)
		server_socket = skt
		server_socket:settimeout(0)
	end)
	if not server_socket or err then
		log("Unable to start TCP server", err)
		return false, err
	end
	return true
end

--- Stop the server. This will close the server socket
-- and close any client connections
function M.stop()
	if server_socket then
		server_socket:close()
		server_socket = nil
	end
	while #clients > 0 do
		local client = table.remove(clients)
		client.socket:close()
	end
end

--- Update the server. The server will listen for new connections
-- and read from connected client sockets.
function M.update()
	if not server_socket then
		return
	end

	-- new connection?
	local client_socket, err = server_socket:accept()
	if client_socket then
		client_socket:settimeout(0)
		local client_ip, client_port = client_socket:getsockname()
		M.handle_client_connected(client_ip, client_port, client_socket)
	end

	-- incoming data?
	for _,client in ipairs(clients) do
		local ok, err = client.reader.receive()
		if not ok then
			log(err)
			M.handle_client_disconnected(client)
		end
	end
end

return M
